File: source/git.ts
import {exec} from 'child_process'
import {promisify} from 'util'
import {existsSync} from 'fs'

const run = promisify(exec)

export async function init(): Promise<void> {
	const result = existsSync('meals/.git') ?
		await gitCommand('pull') :
		await run('git clone -q --depth 1 git@github.com:HAWHHCalendarBot/mensa-data.git meals')

	console.log('git init', result)
}

export async function commitAndPush(): Promise<void> {
	await gitCommand('add -A')
	await tryCommit()

	if (process.env['NODE_ENV'] === 'production') {
		const result = await gitCommand('push -u')
		console.log('git push', result)
	}
}

async function gitCommand(command: string): Promise<{stdout: string; stderr: string}> {
	return run(`git -C meals ${command}`)
}

async function tryCommit(): Promise<void> {
	try {
		await gitCommand('commit -m "update" --no-gpg-sign --author "mensa-crawler <hawhh-mensa-crawler@3t0.de>"')
	} catch (error: unknown) {
		if (typeof error === 'object' && error !== null && 'stdout' in error) {
			// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
			const {stdout} = error as any
			if (typeof stdout === 'string') {
				if (stdout.includes('nothing to commit')) {
					// Everything is fine
					return
				}

				console.error('git commit error', stdout)
			}
		}

		throw error
	}
}

File: src/git.rs
use std::path::Path;
use std::process::Command;

use anyhow::{Context as _, anyhow};

fn command(args: &[&str]) -> anyhow::Result<()> {
    let status = Command::new("git")
        .args(args)
        .current_dir("meals")
        .status()
        .context("failed git command")?;

    if status.success() {
        Ok(())
    } else {
        Err(anyhow!(
            "failed git command. Status code {status}. git {args:?}"
        ))
    }
}

pub fn pull() -> anyhow::Result<()> {
    if Path::new("meals/.git").exists() {
        command(&["pull", "--ff-only"])
    } else {
        let status = Command::new("git")
            .arg("clone")
            .arg("-q")
            .args(["--depth", "1"])
            .arg("git@github.com:HAWHHCalendarBot/mensa-data.git")
            .arg("meals")
            .status()
            .context("failed to clone repo")?;

        if status.success() {
            Ok(())
        } else {
            Err(anyhow!("failed to clone/pull. Status code {status}"))
        }
    }
}

#[expect(dead_code)]
fn push() -> anyhow::Result<()> {
    command(&["push"])
}

fn commit() -> anyhow::Result<()> {
    command(&[
        "commit",
        "-m",
        "update",
        "--no-gpg-sign",
        "--author",
        "mensa-crawler <calendarbot-mensa-crawler@hawhh.de>",
    ])
}

fn add_all() -> anyhow::Result<()> {
    command(&["add", "-A"])
}

pub fn commit_and_push() -> anyhow::Result<()> {
    add_all()?;

    // "nothing to commit" errors. Maybe handle in a better way my checking stdout.
    drop(commit());

    #[cfg(not(debug_assertions))]
    push()?;

    Ok(())
}
